### ISSUE TITLE
Question: python processing error?

### ISSUE THREAD

**Issue by aridlands** [dtarb/TauDEM#110]
Hello:
We are seeing an error message when running any of the TauDem tools within ArcGIS 10.4.
Here's a screen grab of the message from the dialogue window:

![taudem_error_message](https://cloud.githubusercontent.com/assets/22733620/21906829/c87348a8-d8c1-11e6-9e1e-c5a9d487152e.PNG)

Background geoprocessing within ArcGIS is OFF;
TauDEM was installed using the TauDEM 5.3 Complete Windows Installer.
Data frame and all data layers are in the same coordinate system.

Any assistance would be greatly appreciated....thank you.
Peter

**Comment by dtarb** [comment]
It looks like you are trying to read/write the results into a geodatabase.  I do not think the GDAL library we are using to read/write grids can do this, and even if it can, I know that the writing of grids as coded in TauDEM does not allow for output to a geodatabase.  If you move the input file out of a geodatabase and make sure that the output files are written to a location out of a geodatabase using one of the formats we support (.tif, .img) then it should work.

**Comment by aridlands** [comment]
Thank you David for the quick feedback. I did some other tests before sending the help request using standalone DEMs (.tif) sitting within folder on the desktop… And still received the same error. 
I will triple check again and let you know.

Thank you, Peter

**Comment by dtarb** [comment]
It is helpful for debugging to try the command line command, as that separates TauDEM installation and MPI issues from ArcGIS and Python issues.  Try type (or cut and paste) at a command prompt the text following command line: in your screen shot (but with file outside of geodatabase), starting with mpiexec ...

The errors above indicate that the TauDEM command did not run, but I do not know if this is because files being in a geodatabase or something else.

**Comment by aridlands** [comment]
Thank you for the feedback.....We ran the TauDEM command line from a desktop folder containing a .tif elevation raster and received the following error message: 

![taudem_error_message2](https://cloud.githubusercontent.com/assets/22733620/21914771/4117c81a-d8ea-11e6-89ec-006cbe8855b3.PNG)

**Comment by dtarb** [comment]
I have not seen this error before and am not familiar with what libeay32.dll does.  

I did some googling about it as well as searched for it on my system and find that there are many copies/versions of it with different times and date stamps including in the ArcGIS\Desktop 10.4\bin, and Program Files\GDAL folders.  Various internet sources indicate that it is a widely used library related to SSL.  None of this information I found that helpful towards suggesting what your problem is or how to fix it.  I suspect that you have a version of this DLL on your system that is incompatible with the one that is supposed to be used by one of the TauDEM components that is using it.  This may be related to the system path, or 32 bit/64 bit GDAL confusion or multiple conflicting versions of MPI (some other programs install and use MPI too).  All I can suggest is looking on your system for problems like this, uninstalling suspicious software that may be conflicting, uninistalling and reinstalling TauDEM, rebooting etc.  

@pkdash built the TauDEM installer and may have some ideas.

Sorry, not very helpful.  If you are able to track this down, please let me know.

**Comment by aridlands** [comment]
Thank you for the feedback…will definitely look into the issues you raise.  I will also attempt installing TauDEM on another PC and running the scripts via the command line prompt.
Thank you
Peter

**Comment by tsandvik** [comment]
Hi!
I also get a LIBEAY32.dll error message, saying "EC_curve_nist2nid" not located in the library. 
Used the TauDEM 5.3.7 Complete Windows Installer on ArcGIS 10.3.1.
(In case you'd want the info..)
/Tore

**Comment by dtarb** [comment]
Does this happen for any function or just some.  Does this happen when you try run a function from the command prompt or from inside ArcGIS or both.  I have no idea what EC_curve_nist2nid does and am trying to figure out where in the system this may be getting triggered.

**Comment by tsandvik** [comment]
Well, I have only tested the "D8 Flow Directions", "D8 Contributing Area", "Pit Remove" and "Grid Network" tools from the ArcGIS toolset, but they all give error messages quite similar to the one in the original post. The error message described in my previous post was generated from a PitRemove prompt command. 

This TauDEM installation was done on a temporary PC, I'll soon try to install on my ordinary PC, hopefully with more success.

**Comment by dtarb** [comment]
So it seems like all functions and related to C++ executable programs (not ArcGIS python wrappers).

**Comment by chrisnytch** [comment]
Hi, I have been experiencing the same issue originally described by aridlands. I successfully installed TauDEM 5.3.7, and can open the toolbox in ArcGIS, but when trying to execute functions, they fail. This occurs for all tools I try.

![image](https://cloud.githubusercontent.com/assets/25802665/22991975/908fcd42-f394-11e6-9b70-5d66c3b666e5.png)
 
I'm not writing to geodatabases. I've tried running from command prompt and I get the same error about ordinal 4445 and the dynamic link library LIBEAY32.dll. I've tried uninstalling/reinstalling using several previous versions of TauDEM going back to 5.1, and also of ArcMAP, including 10.2.1, 10.2.2, and 10.3.1. 

I'm running windows 7 x64 with service pack 1. On other PCs with the same operating system and software, the functions run. As far as I can tell, I do not have multiple/conflicting versions of GDAL or MPI. So, as dtarb suggested, there's probably something with the DLL version on my machine that is bugging out when TauDEM tries to use it. Would be great to fix this, but no sure how to proceed.

**Comment by dtarb** [comment]
Here are a few stab in the dark suggestions.  Since the error occurs from the command prompt the problem is not Python or ArcMap related.  It may be MPI or GDAL related.  I ran the program Dependency Walker (http://www.dependencywalker.com/) on my system  and found the following 
![image](https://cloud.githubusercontent.com/assets/3456833/23005153/fbdaba1c-f3b7-11e6-8580-5e73c8a9d7dd.png)

I do not know how to interpret all the information there and there is information about it not handling all types of dependencies (https://en.wikipedia.org/wiki/Dependency_Walker). But this does appear to indicate that the GDAL201.dll library that TauDEM uses depends on LIBEAY32.DLL.  On my system both gdal201.dll and libeay32.dll are in the folder C:\Program Files\GDAL.  This should be on the system path, but depending on the order software has been installed may be anywhere on the system path.  You can see the system path by, at a command prompt typing path.  If you type path >temp.txt you get the path saved to a file.  You can edit the path by right clicking on "This PC" on desktop and then "Advanced System Settings", "Environment Variables" and then Edit Path in system variables.  I suggest experimenting with putting C:\Program Files\GDAL at the front of the path so that the GDAL version of libeay32.dll is the first found.  In doing this be careful to record what you do so you can undo if it does not work, as moving things on the path may impact other programs.  You can also use the where command to see where various dll's are in your system.  For example on my system
![image](https://cloud.githubusercontent.com/assets/3456833/23005429/a4e051ca-f3b9-11e6-89aa-99a13b064f27.png)

shows that the GDAL version is actually last found (but evidently the ones before do not interfere or work).  If in your system you see other libeay32.dll files ahead of the GDAL one, changing the path may change this.

Note that every time a path is changed, you need to restart any programs (e.g. the command prompt) to have the change take effect for that program.

**Comment by tsandvik** [comment]
Hello, I've now returned my temporary PC and installed TauDEM on my normal PC. The software seems to be running fluently on this PC. So, I'm sorry I'm not anymore able to check if the stab in the dark suggestions would work. Looking forward to investigate these tools further, though! Thanks!
/Tore

**Comment by chrisnytch** [comment]
THanks for the suggestions, David. I'm playing around with moving the GDAL folder to the front of the path list. One thing I noticed is that during the installation process, there are GDAL files loaded to the C:\Program Files\GDAL, as well as a separate GDAL folder directly under the C:\ directory. Both have a LIBEAY32.DLL file, but in the C:\GDAL folder, it's in a subfolder called "lib". I'm trying different variations of moving the two LIBEAY32.DLL files to different places so they can be "seen" more directly, but still to no avail. A question I have: are there supposed to be two GDAL folders that call applications for TauDEM? I just removed & reloaded TauDEM 5.3.7 and see that there were files installed in both of the GDAL folders.

**Comment by dtarb** [comment]
C:\program files\GDAL holds the C++ GDAL library that the majority of the TauDEM functions use and holds libeay32.dll.

C:\GDAL holds the python GDAL library.   The file in this should be (at least on my system is) libeay32.lib.  Not DLL.  There are only two functions in the SINMAP Stability Index group that are written in python that use this library.  The rest of the functions only use the C++ GDAL library.

Since the problem you are having is with C++ executables I do not think fiddling with C:\GDAL should make any difference and that you should focus on the position in the path of C:\program files\GDAL.

**Comment by chrisnytch** [comment]
Yes, that is correct. I wrote the details of the file extensions incorrectly. Will continue troubleshooting per your suggestion and let you know what I find.

**Comment by chrisnytch** [comment]
Still no luck. I've played around with various path position of C:\program files\GDAL, even eliminating everything else in the path except the GDAL folder, the MPI folder, and the TauDEM folder, but I still get the same error message about missing ordinal 4445. I also searched for other libeay.dll files and found one in the C:\Windows\System32 but there's nothing in the path which calls that file. I'll continuing exploring, but welcome additional suggestions if any come to mind.

**Comment by dtarb** [comment]
A painful trial and error slog installing, uninstalling, rebooting etc is all I can suggest. You could try run dependency walker on your machine and see if it helps you figure out which dll's are involved.  There obviously must be some difference between your computer and the other ones where it works, but tracking this down is a challenge.

**Comment by chrisnytch** [comment]
Thanks for your help. If I figure it out, I'll be sure to post the results for the benefit of others. In the meantime, I'll continue to use TauDEM on computers.
